### PR TITLE
[2025H1] Continue `cargo-semver-checks` goal from 2024H2.

### PR DESCRIPTION
**Pull request by obi1kenobi**
[Rendered](https://github.com/obi1kenobi/rust-project-goals/blob/d26e60f40dd609d1c23dc48bfcf21a6700df9f55/src/2025h1/cargo-semver-checks.md).

Continuation of [the 2024H2 goal](https://rust-lang.github.io/rust-project-goals/2024h2/cargo-semver-checks.html). Check out its tracking issue [here](https://github.com/rust-lang/rust-project-goals/issues/104).

### CODE REVIEW

File: src/2025h1/cargo-semver-checks.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+I (@obi1kenobi) will be working on this effort. The only other resource request would be occasional discussions and moral support from the [cargo] team, of which I already have the privilege as maintainer of a popular cargo plugin.
+
+| Subgoal                                            | Owner(s) or team(s)     | Notes |
+| -------------------------------------------------- | ----------------------- | ----- |
+| Prototype cross-crate linting using workarounds    | @obi1kenobi             |       |
+| Allow linting generic types, lifetimes, bounds     | @obi1kenobi             |       |
+| Handle "special cases" like `'static` and `?Sized` | @obi1kenobi             |       |
+| Handle `#[doc(hidden)]` in sealed trait analysis   | @obi1kenobi             |       |
+| Discussion and moral support                       | ![Team][] [cargo]       |       |
```
Comment (aDotInTheVoid) [inline src/2025h1/cargo-semver-checks.md:145]: I think you can also count on "discussion and moral support" from T-Rustdoc as-well here. You can probably expect small-simple things from us (eg rust-lang/rust#133715, rust-lang/rust#119246), but not larger things (the big one presumably being robust cross-crate Id's (which I don't think has a tracking issue, welp).

Comment (obi1kenobi) [reply src/2025h1/cargo-semver-checks.md:145]: Sounds great, I can add it — thank you.

Any idea if for the purposes of the goal tracking, that should be a separate row in the table, or another team entry in the same row? I know there's some automation around this and I don't want to break it 😅

Comment (obi1kenobi) [inline src/2025h1/cargo-semver-checks.md:145]:
```suggestion
| Discussion and moral support                       | ![Team][] [cargo] [rustdoc] |       |
```

Comment (obi1kenobi) [reply src/2025h1/cargo-semver-checks.md]: Found a precedent for it in another proposal :)

File: src/2025h1/cargo-semver-checks.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+```
+Analogous breaking changes to function return values, struct fields, and associated types would also be missed by `cargo-semver-checks` today.
+
+The main difficulty here lies with the expressiveness of the Rust type system. For example, none of the following changes are breaking:
+```rust
+// previous release:
+pub fn example(value: String) {}
+
+// new release:
+pub fn example(value: impl Into<String>) {}
```
Comment (ogoffart) [inline src/2025h1/cargo-semver-checks.md:105]: Apologies if this is out of topic, but I didn't know where else to report this.

The text is saying that it is not a breaking change to turn a `String` into a `impl Into<String>`, but it is a breaking change when considering type inference.

```rust
example(Default::default());
example("hello".into());
```

Both call worked in the former case but not anymore in the second case.

Comment (obi1kenobi) [reply src/2025h1/cargo-semver-checks.md:105]: Yes, you're right — this is just imprecise use of language on my part.

The changes are not major, in the sense of requiring a SemVer major bump. They are breaking for the current implementation of inference. They are not breaking in an absolute sense, in that better tooling (`rustc` or other tools) could avoid the breakage without changing program semantics. This is why inference-related breakage does not require a major bump in Rust.

More info here: https://predr.ag/blog/some-rust-breaking-changes-do-not-require-major-version/

**Pull request merged**